Mount special filesystem in chroot runners

stagnation opened the pull request:
This mounts '/proc' and '/sys' in the input root of runners if 'chroot' is enabled. It is done with "at"-syscall semantics in the 'Directory', to avoid side effects. A program-scope mutex is required for unmounting because there is no easy-to-use "unmountat".
https://github.com/buildbarn/bb-remote-execution/issues/115

stagnation commented on pkg/runner/local_runner.go:
```diff
@@ -136,6 +157,44 @@ func (r *localRunner) Run(ctx context.Context, request *runner.RunRequest) (*run
+	if requiresSpecialFilesystems {
+		var dir filesystem.Directory
+		dir = r.buildDirectory
+		// TODO(nils): How to best open the input root `Directory`
```
This seems to be a misunderstanding on my part.
I want (to hold on to) a `Directory` object for the input root.
It can be reopened, but do not want to call this opening code twice, I don't think.

EdSchouten replied:
Yeah, you likely also don't want it, because the directory can be moved around to other places, right?

stagnation commented on pkg/runner/local_runner.go:
```diff
@@ -136,6 +157,44 @@ func (r *localRunner) Run(ctx context.Context, request *runner.RunRequest) (*run
+		var dir filesystem.Directory
+		dir = r.buildDirectory
+		// TODO(nils): How to best open the input root `Directory`
+		// It must be a `localDirectory`, but outer directories can be `lazyDirectory`.
```
This strings-split is a cop out for my lack of understanding of the builder and scope walker.
I want to replace this block.

:-1:

EdSchouten replied:
No worries. I can also help out with this part, after we get the bb-storage part sorted out.

stagnation commented on pkg/runner/local_runner.go:
```diff
@@ -166,6 +225,14 @@ func (r *localRunner) Run(ctx context.Context, request *runner.RunRequest) (*run
 		}
 	}
 
+	if requiresSpecialFilesystems {
```
This is not defered, as I saw some unmounting errors during development that would not surface well.
Do you want me to create a defer-lambda that modifes the `err` on the way out to set an error
or is this an okay clean-up section?

EdSchouten replied:
Yeah, having it as a `defer func() {}()` makes sense.

stagnation reviewed (commented):
:-1:

To my reading the mount calls require 'CAP_SYS_ADMIN', there is no 'CAP_MOUNT', so this may have consequences for deployments. But then again, to use `chroot` where the tools crash is also ill-advised.

EdSchouten commented on pkg/runner/local_runner.go:
```diff
@@ -136,6 +157,44 @@ func (r *localRunner) Run(ctx context.Context, request *runner.RunRequest) (*run
+				defer dircloser.Close()
+			}
+
+			dir = dircloser.(filesystem.Directory)
```
Doesn't DirectoryCloser implement Directory?

EdSchouten reviewed (commented):
Thanks for working on this, Nils!

EdSchouten commented on pkg/runner/local_runner.go:
```diff
@@ -136,6 +154,44 @@ func (r *localRunner) Run(ctx context.Context, request *runner.RunRequest) (*run
+		for _, mount := range mounts {
+			err := MountRootDirectory.Mkdir(mount.mountpoint, 0o555)
+			if err != nil {
+				if err != unix.EEXIST {
```
os.IsExist() ?

EdSchouten commented on pkg/runner/local_runner.go:
```diff
@@ -3,20 +3,38 @@ package runner
+		fstype     string
+		source     string
+	}{
+		{procMountpointComponent, "proc", "/proc"},
```
Should this be specified in config?
```proto
repeated MountPointConfiguration mount_points = 123;
```

In main.go:
```go
for _, mountPoint := range configuration.MountPoints {
     runner = NewMountingRunner(runner, mountPoint)
}
```
?

stagnation replied:
Yepp!

EdSchouten reviewed (commented):
General remark: Could this be implemented as a decorator for Runner?

```go
type mountingRunner struct {
     base Runner
}
```

stagnation commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,102 @@
+// that mounts `mount` before running a build action.
+//
+// This decorator can be used for chroot runners
+// that must mount special filesystems into the input root.
```
We could also compare it to the clean runner that has generalized pre- and post-command execution.

stagnation commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,102 @@
+	defer func() {
+		// TODO: is `root` open here or do we need to open it again?
+		if err2 := root.Unmount(mountpoint); err2 != nil {
+			err = util.StatusFromMultiple([]error{
```
is this right? if there was not preceding error we should not have multiple. Maybe the function is helpful and just returns one here, or we should add a check.

stagnation replied:
added a check

EdSchouten commented on pkg/proto/configuration/bb_runner/bb_runner.proto:
```diff
@@ -131,4 +131,23 @@ message ApplicationConfiguration {
+  repeated mount_info input_root_mounts = 15;
+}
+
+message mount_info {
```
Could you please use PascalNaming here? InputMountOptions or something?

EdSchouten commented on pkg/proto/configuration/bb_runner/bb_runner.proto:
```diff
@@ -131,4 +131,23 @@ message ApplicationConfiguration {
   map<string, string> apple_xcode_developer_directories = 14;
+
+  // Mount special filesystems in the input root.
+  // This is useful when running with `chroot`.
```
> This is useful in combination with `chroot_into_input_root`.

Could you please wrap this documentation at ~70-80 columns?

EdSchouten commented on pkg/proto/configuration/bb_runner/bb_runner.proto:
```diff
@@ -131,4 +131,23 @@ message ApplicationConfiguration {
+  // that are created when the operating system boots.
+  // An input root with a full userland implementation may need these.
+  // Typical choices are:
+  //    {"proc", "/proc", "proc"}
```
Please provide full JSON excerpts here, including the keys of the fields.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+		}
+
+		iterator = dir.(filesystem.Directory)
+		if err != nil {
```
This error handling is in the wrong place.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	iterator := buildDirectory
+	var dir filesystem.DirectoryCloser
+
+	parts := strings.Split(inputRootDirectory, "/")
```
Please use `path.Resolve()` here.

```go
if err := path.Resolve(thePath, path. NewRelativeScopeWalker(componentWalker)); err != nil {
     ...
}
```

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	var dir filesystem.DirectoryCloser
+
+	parts := strings.Split(inputRootDirectory, "/")
+	if len(parts) == 0 {
```
The input root directory is in theory permitted to be empty. There is nothing in the runner protocol that forbids it. It may also be ".".

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	}
+	for i, segment := range parts {
+		var err error
+		component := path.MustNewComponent(segment)
```
This should use a regular `NewComponent()`. You can't assume the provided path is well-formed.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	}
+	defer root.Close()
+
+	mountpoint := path.MustNewComponent(r.mount.Mountpoint)
```
What about mounts in some nested location?

stagnation replied:
Good point! They could be in any directory hierarchy, the code does not support that as it is written yet.

Aside, nested mounts: the mounts themselves can also be nested, If one wants to use the original filesystem as a mountpoint the innermost path must be given first.

```
mkdir -p outer/inner/x
mount -t sysfs /sys outer/inner
mount -t sysfs /sys outer
mount -t sysfs /sys outer/inner/x  # x is not inside the mounted sysfs
mount -t sysfs /sys outer/inner/block # can be mounted ontop of the other sysfs.
```

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	defer root.Close()
+
+	mountpoint := path.MustNewComponent(r.mount.Mountpoint)
+	err = root.Mkdir(mountpoint, 0o555)
```
Should we be creating the directories ourselves? Maybe it makes more sense to only establish these mounts if they actually exist in the input root?

stagnation replied:
Good point, with this as an advanced feature, we can probably require very strict compliance from clients that wants to use it. I can require that in the proto specification.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+		return nil, util.StatusWrapf(err, "Failed to mount '%#v' in the input root", r.mount)
+	}
+	defer func() {
+		if err2 := root.Unmount(mountpoint); err2 != nil {
```
With only the `Run()` call in between, I would not recommend using `defer` here. Just place this code right after calling `Run()`.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,105 @@
+	err = root.Mkdir(mountpoint, 0o555)
+	if err != nil {
+		if ! os.IsExist(err) {
+			return nil, util.StatusWrapf(err, "Failed to create mount point: '%#v' in the input root", r.mount.Mountpoint)
```
There is no need to put quotes around strings if you're using `%#v`. It already does that for you.

stagnation committed:
Fix typo

stagnation committed:
Mount special filesystems in chroot runners

This can be used to mount special filesystems like '/proc' and '/sys' in
the input root of actions if 'chroot' is enabled. The filesystems are
required for many tools to work.

Solves: #115

EdSchouten commented on cmd/bb_runner/main.go:
```diff
@@ -68,6 +68,14 @@ func main() {
 			buildDirectoryPath,
 			commandCreator,
 			configuration.SetTmpdirEnvironmentVariable)
+		for _, mountinfo := range configuration.InputRootMounts {
```
I actually prefer the way the code is written right now. Nothing wrong with looping over an empty list.

EdSchouten commented on pkg/proto/configuration/bb_runner/bb_runner.proto:
```diff
@@ -131,4 +131,37 @@ message ApplicationConfiguration {
   map<string, string> apple_xcode_developer_directories = 14;
+
+  // Mount special filesystems in the input root. This is useful when
+  // running with `chroot_into_input_root`. some tools require access to
```
s/some/Some/

stagnation replied:
:eyes: or my favorite metacharacter: s/some/\u&/

EdSchouten commented on pkg/proto/configuration/bb_runner/bb_runner.proto:
```diff
@@ -131,4 +131,37 @@ message ApplicationConfiguration {
+message InputMountOptions {
+  // Mount a filesystem in the input root, a relative path.
+  string mountpoint = 1;
+  // Source filesystem from the runner's operating system, absolute path.
```
I wouldn't document this as being an absolute path. What this string needs to be is actually specific on the type of file system used. In the case of NFS it would be of the form 192.168.1.1:/foo.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,75 @@
+	scopeWalker := path.NewRelativeScopeWalker(&pathResolver)
+	defer pathResolver.closeAll()
+
+	fullpath := filepath.Join(request.InputRootDirectory, r.mount.Mountpoint)
```
My suggestion would be to not use `filepath.Join()` here. It assumes that the native path separator is used, even though `request.InputRootDirectory` always uses forward slashes, even on Windows. There is also no need to do that if you just call `path.Resolve()` multiple times:
```go
if err := path.Resolve(request.InputRootDirectory, path.NewRelativeScopeWalker(&pathResolver)); err != nil {
    ...
}
if err := path.Resolve(r.mount.Mountpoint, path.NewRelativeScopeWalker(&pathResolver)); err != nil {
    ...
}
```

I would also strongly advise against using the same PathResolver for both processing the input root directory and the mount point. Notice how the path resolver contains a stack. That stack is there to process `".."` entries. Because the stack isn't cleared/truncated, a mount point containing `".."` components might leave the input root directory.

This is not a big issue right now, because the mount point is fully specified in config. But if someone ever comes along and wants the mount point path resolution to respect symlinks, it can become dangerous fairly quickly.

EdSchouten commented on pkg/runner/mounting_runner.go:
```diff
@@ -0,0 +1,75 @@
+		stack: util.NewNonEmptyStack(filesystem.NopDirectoryCloser(r.buildDirectory)),
+	}
+	scopeWalker := path.NewRelativeScopeWalker(&pathResolver)
+	defer pathResolver.closeAll()
```
This might keep more file descriptors open than strictly necessary. Would there be a way for us to close intermediate directories while `Run()` is called?

EdSchouten reviewed (commented):
Thanks for working on this!

stagnation commented:
I think this is a flaky storage error in the check, but other storage error that occurred around the same time had more explicit error messages.
```
ERROR: /home/runner/work/bb-remote-execution/bb-remote-execution/cmd/bb_scheduler/BUILD.bazel:93:9: ImageLayer cmd/bb_scheduler/bb_scheduler_container-layer.tar failed: (Exit 1): build_tar failed: error executing command (from target //cmd/bb_scheduler:bb_scheduler_container) bazel-out/k8-opt-exec-2B5CBBC6-ST-625e526ca8a8/bin/external/io_bazel_rules_docker/container/build_tar ... (remaining 5 arguments skipped)
```

stagnation commented:
With https://github.com/buildbarn/bb-remote-execution/pull/225 we'll have more reason to use this :)

moroten commented:
Instead of `unmountat`, would a mount namespace help?
